applicationset/services/scm_provider/aws_codecommit.go

```diff
@@ -0,0 +1,371 @@
+type AWSTaggingClient interface {
+	GetResourcesWithContext(aws.Context, *resourcegroupstaggingapi.GetResourcesInput, ...request.Option) (*resourcegroupstaggingapi.GetResourcesOutput, error)
+}
+
+type AWSCodeCommitProvider struct {
+	codeCommitClient AWSCodeCommitClient
+	taggingClient    AWSTaggingClient
+	tagFilters       []*application.TagFilter
+	allBranches      bool
+}
+
+func NewAWSCodeCommitProvider(ctx context.Context, tagFilters []*application.TagFilter, role string, region string, allBranches bool) (*AWSCodeCommitProvider, error) {
+	taggingClient, codeCommitClient, err := createAWSDiscoveryClients(ctx, role, region)
+	if err != nil {
+		return nil, err
+	}
+	return &AWSCodeCommitProvider{
+		codeCommitClient: codeCommitClient,
+		taggingClient:    taggingClient,
+		tagFilters:       tagFilters,
+		allBranches:      allBranches,
+	}, nil
+}
+
+func (p *AWSCodeCommitProvider) ListRepos(ctx context.Context, cloneProtocol string) ([]*Repository, error) {
+	repos := make([]*Repository, 0)
+
+	repoNames, err := p.listRepoNames(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list codecommit repository due to %w", err)
+	}
+
+	for _, repoName := range repoNames {
+		repo, err := p.codeCommitClient.GetRepositoryWithContext(ctx, &codecommit.GetRepositoryInput{
+			RepositoryName: aws.String(repoName),
+		})
+		if err != nil {
+			// we don't want to skip at this point. It's a valid repo, we don't want to have flapping Application on an AWS outage.
+			return nil, fmt.Errorf("failed to get codecommit repository due to %w", err)
+		}
+		if repo.RepositoryMetadata.DefaultBranch == nil || *repo.RepositoryMetadata.DefaultBranch == "" {
+			// if a codecommit repo doesn't have default branch, it's uninitialized. not going to bother with it.
+			log.Warnf("repository %s does not have default branch, skipped", repoName)
+			continue
+		}
+		var url string
+		switch cloneProtocol {
+		// default to SSH if unspecified (i.e. if "").
+		case "", "ssh":
+			url = *repo.RepositoryMetadata.CloneUrlSsh
+		case "https":
+			url = *repo.RepositoryMetadata.CloneUrlHttp
+		case "https-fips":
+			url, err = getCodeCommitFIPSEndpoint(*repo.RepositoryMetadata.CloneUrlHttp)
+			if err != nil {
+				return nil, fmt.Errorf("https-fips is provided but repoUrl can't be transformed to FIPS endpoint. %w", err)
+			}
+		default:
+			return nil, fmt.Errorf("unknown clone protocol for codecommit %v", cloneProtocol)
+		}
+		repos = append(repos, &Repository{
+			// there's no "organization" level at codecommit.
+			// we are just using AWS accountId for now.
+			Organization: *repo.RepositoryMetadata.AccountId,
+			Repository:   *repo.RepositoryMetadata.RepositoryName,
+			URL:          url,
+			Branch:       *repo.RepositoryMetadata.DefaultBranch,
+			// we could propagate repo tag keys, but without value not sure if it's any useful.
+			Labels:       []string{},
+			RepositoryId: *repo.RepositoryMetadata.RepositoryId,
```

Comment:
Should we protect these pointer dereferences with nil checks?

Reply:
Good point, I've added some more defensive checks, and replaced all pointer dereferences with `aws.StringValue()`.